### ISSUE TITLE
Port the loss scale optimizer to keras-core

### ISSUE THREAD

**Issue by mattdangerw** [keras-team/keras-core#571]
We have been seeing some bad training runs in keras-nlp with a LLM running under mixed precision, and I think at least one of the missing pieces is scaling the loss. Scaling the loss up by 1k appeared to bring it in line with full precision.

https://github.com/keras-team/keras/blob/master/keras/mixed_precision/loss_scale_optimizer.py

This is a good explainer for why we need it ->
https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html#training

**Comment by mattdangerw** [comment]
@fchollet does adding this make sense to you?

Also any thoughts on the general approach, does an optimizer wrapper that takes an inner optimizer still sound like the approach we should go with?

**Comment by fchollet** [comment]
Do you think this is something we should do as a base class optimizer argument, or as a separate wrapper class (like we had before)? For backwards compat, perhaps it's best if we just bring back the class.

**Comment by chenmoneygithub** [comment]
The TF LossScaleOptimizer is pretty complex ([link](https://github.com/keras-team/keras/blob/v2.13.1/keras/mixed_precision/loss_scale_optimizer.py#L1085-L1434)), so I feel a separate class would be better. Also I feel the complexity could be reduced in the new implementation.

**Comment by mattdangerw** [comment]
Definitely removing complexity would be great! Current implementation has a lot going on, metaclasses, etc.

In keras-core, current optimizer classes do not deal with `loss` directly. Would we need to change the optimizer interface to allow loss scaling? And if so, should we handle this in `Trainer` instead of `Optimizer`?

In terms of exposing, part of me was wondering if we could do something simpler, like an option. E.g. `loss_scale="auto"` (dynamic) or `loss_scale=128.` (fixed). But if we do want to support the full set of options currently provided by `LossScaleOptimizer` (setting how often the scale is adjusted, setting the initial scale), then a class probably does make sense.

First thing could be to make some sort of prototype, and see if this requires changes to our optimizer/trainer interfaces.

**Comment by fchollet** [comment]
> In keras-core, current optimizer classes do not deal with loss directly. Would we need to change the optimizer interface to allow loss scaling?

Presumably you can just override `apply` and scale up the gradients? That should be the same effect as scaling up the loss?

If you really need to scale up the loss itself for whatever reason, we'd need to make it a `Model` option applied in `compute_loss`.

Either way it sounds like 3 lines of code?

**Comment by mattdangerw** [comment]
> Presumably you can just override apply and scale up the gradients? That should be the same effect as scaling up the loss?

I don't think so. I think the important thing is to scale up the loss **before** backprop. In tf, you are scaling up the loss inside the tape. By blowing up the loss to a larger range of values, before backprop, it means your gradients can avoid getting "washed out" to all zeros due to loss of precision. I think `washed_out_gradients * loss_scale` would still be washed out. Not really sure, that was my quick read of it today.

I think roughly what we need in tensorflow pseudocode is this...

```python
def train_step(self, data):
    x, y, sample_weight = unpack_x_y_sample_weight(data)

    with tf.GradientTape() as tape:
        y_pred = self(x, training=True)
        loss = self.compute_loss(x=x, y=y, y_pred=y_pred, sample_weight=sample_weight)
        # NEW: scale the loss *before* computing gradients.
        loss = loss * self.scale

    trainable_weights = [v.value for v in self.trainable_weights]
    gradients = tape.gradient(loss, trainable_weights)

    # NEW: update the loss scale. Only apply gradients if they are finite, otherwise skip.
    if is_fininte(gradients):
        self.counter += 1
        if self.counter > self.update_frequency:
            self.counter = 0
            self.scale *= 2
        self.optimizer.apply_gradients(zip(gradients, trainable_weights))
    else:
        self.counter = 0
        self.scale /= 2

    return self.compute_metrics(x, y, y_pred, sample_weight=sample_weight)
```

We need a new variable for a counter. We need to scale loss before computing gradients. And we need to check our gradients to figure out how to update our dynamic scale each step.

**Comment by fchollet** [comment]
Taking into account backwards compatibility, I think we should probably bring back `keras.mixed_precision.LossScaleOptimizer`? We can always have `model.compute_loss` check for a `loss_scale` parameter on the optimizer to optional scale the loss before computing gradients -- though that's a big too spaghetti for my taste.

**Comment by mattdangerw** [comment]
Makes sense to me!

I do think that will require a minor tweak to the optimizer interface. Either bring back something like the `compue_gradients(loss, var_list, tape)` in `tf.keras` (this is how `LossScaleOptimizer` works today). Or add something much simpler method to the interface like `loss = optimizer.get_scaled_loss(loss)` and stick that in `compute_loss`. The latter seems simpler to me.

Commented the tf train step below...

```python
    def train_step(self, data):
        x, y, sample_weight = data_adapter_utils.unpack_x_y_sample_weight(data)

        with tf.GradientTape() as tape:
            y_pred = self(x, training=True)
            loss = self.compute_loss(
                x=x, y=y, y_pred=y_pred, sample_weight=sample_weight
            ) # Loss scaled HERE.

        if self.trainable_weights:
            trainable_weights = [v.value for v in self.trainable_weights]
            gradients = tape.gradient(loss, trainable_weights)
            self.optimizer.apply_gradients(zip(gradients, trainable_weights)) # The rest handled HERE.

        return self.compute_metrics(x, y, y_pred, sample_weight=sample_weight)
```

**Comment by fchollet** [comment]
Maybe we could do it like this:

- Base `Optimizer` has a `scale_loss` method and a `loss_scale_factor=1.0` argument.
- `Trainer` has a `scale_loss` method which defaults to `self.optimizer.scale_loss` if `self.optimizer` exists and is a Keras optimizer.
- `keras.mixed_precision.LossScaleOptimizer` is added as a lightweight wrapper for backwards compat only


Typical usage would then be `model.compile(..., optimizer=AdamW(loss_scale_factor=8.0))`.

What do you think?

**Comment by mattdangerw** [comment]
Two things...

1. I think the 90% use case here is not constant scaling the loss, but auto scaling the loss. Probably `LossScaleOptimizer` can be the place that lives. I think the `scale_loss()` setup you propose works, but in the case of `LossScaleOptimizer` the scale will be dynamically updated, based on the presence of `nan`s in the gradient during `apply()`.
2. I think we need to record the "unscaled loss" in the `_loss_tracker`, otherwise the reported loss will scare everyone off. But we need to scale the loss inside the `GradientTape`/`jax.grad` function. That might require some trainer specific changes. E.g. move the `_loss_tracker` update in jax inside the `compute_loss_and_updates` before scaling. And in tf move the `_loss_tracker` update inside the tape. Would that work?

Iterating on the proposal...

- Base `Optimizer` has a `scale_loss()` method and a `loss_scale_factor=1.0` argument.
- `Trainer` has a `scale_loss()` method which defaults to `self.optimizer.scale_loss` if `self.optimizer` exists and is a Keras optimizer.
- The method is called inside `train_step` after updating the `_loss_tracker` (backend specific).
- `keras.mixed_precision.LossScaleOptimizer` will update an internal scale variable up and down during `apply`, and use that to change the `scale_loss()` output.
- If the precision policy is mixed, `Trainer` will automatically wrap the passed optimizer in a `LossScaleOptimizer` with default options (should we add a toggle for this?)

What about this?

**Comment by fchollet** [comment]
SGTM!

**Comment by AlexanderLavelle** [comment]
Just to ensure I understand -- this will now behave as intended without the additional lines of code i.e.:

```python
optimizer = keras.optimizers.Adam()
- optimizer = keras.mixed_precision.LossScaleOptimizer(optimizer)
model.compile(optimizer=optimizer)
```

and then 
<inside train_step>
```python
loss = loss_fn( ... )
- scaled_loss = optimizer.get_scaled_loss(loss)
- scaled_grads = tape.gradient(scaled_loss, trainable_weights)
- gradients = optimizer.get_unscaled_gradients(scaled_grads)

+ gradients = tape.gradient(loss, trainable_weights)
optimizer.apply_gradients(zip(gradients, trainable_weights))
```

and in the meantime, scaled_loss can be multiplied by 1,000?

**Comment by james77777778** [comment]
Hi @mattdangerw @fchollet 

Is there any update on this issue?
Loss scaling plays a crucial role in mixed precision, which is important for training large models.

Would a contribution be appreciated? If yes, I'm willing to work on the PR.
Thanks!

**Comment by mattdangerw** [comment]
@james77777778 sorry for the delayed reply here! Missed this.

I started working on this literally the day before you messaged (when I assigned myself). But should have been checking for messages more! We may end up going with the version I put up because it is backward compatible with `tf.keras`. E.g. it gives a compatible signature with https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/LossScaleOptimizer

But very much appreciate the contribution nonetheless!
